experiment/simulation/js/layout.js

```diff
@@ -0,0 +1,161 @@
+import { simulate, deleteElement } from "./gate.js";
+import { connectGate, connectDecoderDisplay, unbindEvent, initDecoder, init7Segment, refreshWorkingArea } from "./main.js";
+import { simulateDecoder } from "./decoder.js";
+"use strict";
+
+// Wires
+export const wireColours = [
+  "#ff0000",
+  "#00ff00",
+  "#0000ff",
+  "#bf6be3",
+  "#ff00ff",
+  "#00ffff",
+  "#ff8000",
+  "#00ff80",
+  "#80ff00",
+  "#ff0080",
+  "#8080ff",
+  "#c0c0c0",
+];
+
+// Contextmenu
+const menu = document.querySelector(".menu");
+const menuOption = document.querySelector(".menu-option");
+let menuVisible = false;
+
+const toggleMenu = command => {
+  menu.style.display = command === "show" ? "block" : "none";
+  menuVisible = !menuVisible;
+};
+
+export const setPosition = ({ top, left }) => {
+  menu.style.left = `${left}px`;
+  menu.style.top = `${top}px`;
+  toggleMenu("show");
+};
+
+window.addEventListener("click", e => {
+  if (menuVisible) toggleMenu("hide");
+  window.selectedComponent = null;
+  window.componentType = null;
+});
+
+menuOption.addEventListener("click", e => {
+  if (e.target.innerHTML === "Delete") {
+    if (window.componentType === "gate") {
+      deleteElement(window.selectedComponent);
+    }
+  }
+  window.selectedComponent = null;
+  window.componentType = null;
+});
+
+// Tabs
+
+function changeTabs(e) {
+  const task = e.target.parentNode.id;
+  if (window.currentTab === task) {
+    return;
+  }
+
+  if (window.currentTab != null) {
+    document.getElementById(window.currentTab).classList.remove("is-active");
+  }
+  window.currentTab = task;
+  document.getElementById(task).classList.add("is-active");
+
+  // Half adder
+  if (task === "task1") {

```

Comment:
Generally every conditional should have a default else case so we know when a conditional reached an unexpected state.

Reply:
the default else will be blank as task can have no other value.

Reply:
Exactly, that is the idea. If the default else is not supposed to happen, when it happens, the system is in an inconsistent state. Catching that error explicitly is better than not doing anything on it